clients/go/admin/token_source_provider.go

```diff
@@ -0,0 +1,120 @@
+package admin
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/flyteorg/flyteidl/clients/go/admin/externalprocess"
+
+	"github.com/flyteorg/flyteidl/clients/go/admin/pkce"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+	"golang.org/x/oauth2"
+)
+
+// TokenSourceProvider defines the interface needed to provide a TokenSource that is used to
+// create a client with authentication enabled.
+type TokenSourceProvider interface {
+	GetTokenSource(ctx context.Context) (oauth2.TokenSource, error)
+}
+
+func NewTokenSourceProvider(ctx context.Context, cfg *Config, tokenCache pkce.TokenCache,
+	authClient service.AuthMetadataServiceClient) (TokenSourceProvider, error) {
+
+	tokenURL := cfg.TokenURL
+	if len(tokenURL) == 0 {
+		metadata, err := authClient.GetOAuth2Metadata(ctx, &service.OAuth2MetadataRequest{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch auth metadata. Error: %v", err)
+		}
+
+		tokenURL = metadata.TokenEndpoint
+	}
+
+	clientMetadata, err := authClient.GetPublicClientConfig(ctx, &service.PublicClientAuthConfigRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch client metadata. Error: %v", err)
+	}
```

Comment:
I think you can safely move this into `AuthTypeClientSecret` case... avoid the extra call when not needed...